core/src/sumeragi/message.rs

```diff
@@ -38,74 +38,126 @@ impl ControlFlowMessage {
 
 /// `BlockCreated` message structure.
 #[derive(Debug, Clone, Decode, Encode)]
-#[non_exhaustive]
 pub struct BlockCreated {
     /// The corresponding block.
     pub block: SignedBlock,
 }
 
-impl From<ValidBlock> for BlockCreated {
-    fn from(block: ValidBlock) -> Self {
+impl From<&ValidBlock> for BlockCreated {
+    fn from(block: &ValidBlock) -> Self {
         Self {
-            block: block.into(),
+            // TODO: Redundant clone
+            block: block.clone().into(),
         }
     }
 }
 
 /// `BlockSigned` message structure.
 #[derive(Debug, Clone, Decode, Encode)]
-#[non_exhaustive]
 pub struct BlockSigned {
     /// Hash of the block being signed.
     pub hash: HashOf<BlockPayload>,
-    /// Set of signatures.
-    pub signatures: SignaturesOf<BlockPayload>,
+    /// Signature of the block
+    pub signature: BlockSignature,
 }
 
 impl From<&ValidBlock> for BlockSigned {
     fn from(block: &ValidBlock) -> Self {
-        let block_hash = block.as_ref().hash_of_payload();
-        let block_signatures = block.as_ref().signatures().clone();
-
         Self {
-            hash: block_hash,
-            signatures: block_signatures,
+            hash: block.as_ref().hash_of_payload(),
+            signature: block
+                .as_ref()
+                .signatures()
+                .last()
+                .cloned()
+                .expect("INTERNAL BUG: Block must have at least one signature"),
         }
     }
 }
 
 /// `BlockCommitted` message structure.
-#[derive(Debug, Clone, Decode, Encode)]
-#[non_exhaustive]
+#[derive(Debug, Clone, Encode)]
 pub struct BlockCommitted {
     /// Hash of the block being signed.
     pub hash: HashOf<SignedBlock>,
```

Comment:
Not this PR issue, but shouldn't `hash` in block messages be unified to `HashOf<BlockPayload>`?

Reply:
let's discuss in #4641